### ISSUE TITLE
If the redis server is unavailble module omhiredis failed with Segmentation fault

### ISSUE THREAD

**Issue by missnebun** [rsyslog/rsyslog#727]
#my test config

module(load="imtcp") 
module(load="imudp")
module(load="omhiredis")

input(type="imudp" port="514" ruleset="remote")
input(type="imtcp" port="514" ruleset="remote")

ruleset(name="remote"
    queue.type="FixedArray"        # we'll have a fixed memory queue for this ruleset
    queue.spoolDirectory="/var/spool/rsyslog" #Spool directory
    queue.filename="remote-input"
    queue.maxdiskspace="10m"
    queue.maxfilesize="5m"
    queue.highwatermark="800" #Start spilling to disk at this size
    queue.lowwatermark="400"   #stop spilling when it gets back to this size
    queue.size="120" # capacity of the queue
    queue.saveonshutdown="on"
    queue.workerthreads="1"
    queue.dequeuebatchsize="100"
    )
  {
    action(
      type="omhiredis"
      name="redis2050-fwd"
      server="10.250.24.205"
      mode="queue"         
      key="rsyslog_logstash"
      )
    action(
      type="omhiredis"
      name="redis2093-fwd"
      server="10.250.24.71"
      mode="queue"
      key="rsyslog_logstash"
	  action.execOnlyWhenPreviousIsSuspended="on"
      )
  }

  
  
  
  2732.137948661:imtcp.c        : epoll returned 1 entries
2732.137981398:imtcp.c        : epoll push ppusr[0]: 0x7f25a06fee90
2732.137996289:imtcp.c        : tcpsrv: ready to process 1 event entries
2732.138007097:imtcp.c        : tcpsrv: processing item 0, pUsr 0x7f25a06fee90, bAbortConn
2732.138097541:imtcp.c        : New connect on NSD 0x7f25a06ff120.
2732.138143215:imtcp.c        : dnscache: entry (nil) found
2732.141204500:imtcp.c        : adding nsdpoll entry 0/0x7f258c003550, sock 12
2732.141232063:imtcp.c        : New session created with NSD 0x7f258c003550.
2732.141243401:imtcp.c        : doing epoll_wait for max 128 events
2732.141255427:imtcp.c        : epoll returned 1 entries
2732.141264062:imtcp.c        : epoll push ppusr[0]: 0x7f258c003550
2732.141273041:imtcp.c        : tcpsrv: ready to process 1 event entries
2732.141281788:imtcp.c        : tcpsrv: processing item 0, pUsr 0x7f258c003550, bAbortConn
2732.141318764:imtcp.c        : netstream 0x7f258c003820 with new data
2732.141381399:imtcp.c        : msg parser: flags 30, from 'nyvm1122', msg '<46>Jan 11 09:31:58 nyvm1122 rsyslogd-2359: action 'action 0'
2732.141395683:imtcp.c        : parse using parser list 0x7f25a06f21f0 (the default list).
2732.141408800:imtcp.c        : Parser 'rsyslog.rfc5424' returned -2160
2732.141427289:imtcp.c        : Message will now be parsed by the legacy syslog parser (one size fits all... ;)).
2732.141442354:imtcp.c        : Parser 'rsyslog.rfc3164' returned 0
2732.141458446:imtcp.c        : msg parser: flags 30, from 'nyvm1122', msg '<46>Jan 11 09:31:58 nyvm1122 rsyslogd-2359: action 'action 0'
2732.141469436:imtcp.c        : parse using parser list 0x7f25a06f21f0 (the default list).
2732.141480099:imtcp.c        : Parser 'rsyslog.rfc5424' returned -2160
2732.141488122:imtcp.c        : Message will now be parsed by the legacy syslog parser (one size fits all... ;)).
2732.141498588:imtcp.c        : Parser 'rsyslog.rfc3164' returned 0
2732.141513438:imtcp.c        : remote: qqueueAdd: entry added, size now log 1, phys 1 entries
2732.141523743:imtcp.c        : remote: qqueueAdd: entry added, size now log 2, phys 2 entries
2732.141535160:imtcp.c        : remote:Reg: high activity - starting 1 additional worker thread(s).
2732.141653266:imtcp.c        : remote:Reg: started with state 0, num workers now 1
2732.141665868:imtcp.c        : remote: MultiEnqObj advised worker start
2732.141678861:imtcp.c        : doing epoll_wait for max 128 events
2732.141728403:7f2583de5700: thread created, tid 15025, name 'rs:remote:Reg'
2732.142951423:remote:Reg/w0  : wti 0x7f25a0710f10: worker starting
2732.142972739:remote:Reg/w0  : DeleteProcessedBatch: we deleted 0 objects and enqueued 0 objects
2732.142985808:remote:Reg/w0  : doDeleteBatch: delete batch from store, new sizes: log 2, phys 2
2732.143000920:remote:Reg/w0  : processBATCH: batch of 2 elements must be processed
2732.143011612:remote:Reg/w0  : processBATCH: next msg 0: <46>Jan 11 09:31:58 nyvm1122 rsyslogd-2359: action 'action 0' resumed (module 'builtin:omfwd') [v8.15.0 try http://www.rsyslog.c
2732.143063751:remote:Reg/w0  :     ACTION 8 [omhiredis:action(type="omhiredis" ...)]
2732.143089033:remote:Reg/w0  : executing action 8
2732.143101288:remote:Reg/w0  : action 'redis2050-fwd': called, logging to omhiredis (susp 0/0, direct q 1)
2732.143125567:remote:Reg/w0  : action 'redis2050-fwd': is transactional - executing in commit phase
2732.143139409:remote:Reg/w0  : wti 0x7f25a0710f10: we need to create a new action worker instance for action 8
2732.143152381:remote:Reg/w0  : DDDD: writing data to table spot 0
2732.143162567:remote:Reg/w0  : wti 0x7f25a0710f10: created action worker instance 1 for action 8
2732.143172686:remote:Reg/w0  : omhiredis: beginTransaction called
2732.143182598:remote:Reg/w0  : Action 8 transitioned to state: itx
2732.143192187:remote:Reg/w0  : action 'redis2050-fwd': set suspended state to 0
2732.143200806:remote:Reg/w0  :     ACTION 9 [omhiredis:action(type="omhiredis" ...)]
2732.143218748:remote:Reg/w0  : executing action 9
2732.143228708:remote:Reg/w0  : action 'redis2093-fwd': called, logging to omhiredis (susp 1/0, direct q 1)
2732.143238363:remote:Reg/w0  : action 'redis2093-fwd': NOT executing, as previous action was not suspended
2732.143249392:remote:Reg/w0  : processBATCH: next msg 1: <46>Jan 11 09:31:58 nyvm1122 rsyslogd-2359: action 'action 0' resumed (module 'builtin:omfwd') [v8.15.0 try http://www.rsyslog.c
2732.143260715:remote:Reg/w0  :     ACTION 8 [omhiredis:action(type="omhiredis" ...)]
2732.143278502:remote:Reg/w0  : executing action 8
2732.143288006:remote:Reg/w0  : action 'redis2050-fwd': called, logging to omhiredis (susp 0/0, direct q 1)
2732.143299844:remote:Reg/w0  : action 'redis2050-fwd': is transactional - executing in commit phase
2732.143310473:remote:Reg/w0  : action 'redis2050-fwd': set suspended state to 0
2732.143319421:remote:Reg/w0  :     ACTION 9 [omhiredis:action(type="omhiredis" ...)]
2732.143336870:remote:Reg/w0  : executing action 9
2732.143346520:remote:Reg/w0  : action 'redis2093-fwd': called, logging to omhiredis (susp 1/0, direct q 1)
2732.143356285:remote:Reg/w0  : action 'redis2093-fwd': NOT executing, as previous action was not suspended
2732.143366025:remote:Reg/w0  : END batch execution phase, entering to commit phase
2732.143379932:remote:Reg/w0  : actionCommitAll: action 8, state 1, nbr to commit 0 isTransactional 1
2732.143390340:remote:Reg/w0  : doTransaction: action 8, currIParam 2
2732.143400332:remote:Reg/w0  : entering actionCalldoAction(), state: itx, actionNbr 8
2732.143411038:remote:Reg/w0  : omhiredis: trying connect to '10.250.24.205' at port 6379
2732.144931309:remote:Reg/w0  : Called LogMsg, msg: can not initialize redis handle
2732.144984945:remote:Reg/w0  : main Q: qqueueAdd: entry added, size now log 1, phys 1 entries
2732.145034257:main Q:Reg/w0  : wti 0x7f25a0717080: worker awoke from idle processing
2732.145047777:main Q:Reg/w0  : DeleteProcessedBatch: we deleted 0 objects and enqueued 0 objects
2732.145057960:main Q:Reg/w0  : doDeleteBatch: delete batch from store, new sizes: log 1, phys 1
2732.145070211:main Q:Reg/w0  : processBATCH: batch of 1 elements must be processed
2732.145080049:main Q:Reg/w0  : processBATCH: next msg 0: can not initialize redis handle [v8.15.0 try http://www.rsyslog.com/e/2007 ]
2732.145092013:main Q:Reg/w0  :     PRIFILT '*.info;mail.none;authpriv.none;cron.none'
2732.145187660:main Q:Reg/w0  :     pmask: 7F 7F  X 7F 7F 7F 7F 7F 7F  X  X 7F 7F 7F 7F 7F 7F 7F 7F 7F 7F 7F 7F 7F 7F 7F
2732.145301943:main Q:Reg/w0  : PRIFILT condition result is 1
2732.145311601:main Q:Reg/w0  :     ACTION 0 [builtin:omfile:/var/log/messages]
2732.145331723:main Q:Reg/w0  : executing action 0
2732.145342851:main Q:Reg/w0  : action 'action 0': called, logging to builtin:omfile (susp 0/0, direct q 1)
2732.145360307:main Q:Reg/w0  : action 'action 0': is transactional - executing in commit phase
2732.145373407:main Q:Reg/w0  : Action 0 transitioned to state: itx
2732.145383930:main Q:Reg/w0  : action 'action 0': set suspended state to 0
2732.145393629:main Q:Reg/w0  :     PRIFILT 'authpriv.*'
2732.145409293:main Q:Reg/w0  :     pmask:  X  X  X  X  X  X  X  X  X  X FF  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X
2732.145512454:main Q:Reg/w0  : PRIFILT condition result is 0
2732.145521279:main Q:Reg/w0  :     PRIFILT 'authpriv.*'
2732.145536583:main Q:Reg/w0  :     pmask:  X  X  X  X  X  X  X  X  X  X FF  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X
2732.145639390:main Q:Reg/w0  : PRIFILT condition result is 0
2732.145648010:main Q:Reg/w0  :     PRIFILT 'mail.*'
2732.145663281:main Q:Reg/w0  :     pmask:  X  X FF  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X
2732.145830778:main Q:Reg/w0  : PRIFILT condition result is 0
2732.145839948:main Q:Reg/w0  :     PRIFILT 'cron.*'
2732.145855125:main Q:Reg/w0  :     pmask:  X  X  X  X  X  X  X  X  X FF  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X
2732.145958386:main Q:Reg/w0  : PRIFILT condition result is 0
2732.145967111:main Q:Reg/w0  :     PRIFILT '*.emerg'
2732.145982078:main Q:Reg/w0  :     pmask:  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1
2732.146094898:main Q:Reg/w0  : PRIFILT condition result is 0
2732.146103791:main Q:Reg/w0  :     PRIFILT 'uucp,news.crit'
2732.146119389:main Q:Reg/w0  :     pmask:  X  X  X  X  X  X  X  7  7  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X
2732.146223325:main Q:Reg/w0  : PRIFILT condition result is 0
2732.146232076:main Q:Reg/w0  :     PRIFILT 'local7.*'
2732.146247118:main Q:Reg/w0  :     pmask:  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X FF  X  X
2732.146350669:main Q:Reg/w0  : PRIFILT condition result is 0
2732.146359576:main Q:Reg/w0  : END batch execution phase, entering to commit phase
2732.146371303:main Q:Reg/w0  : actionCommitAll: action 0, state 1, nbr to commit 1 isTransactional 1
2732.146383439:main Q:Reg/w0  : doTransaction: have commitTransaction IF, using that, pWrkrInfo 0x7f25a0717150
2732.146395724:main Q:Reg/w0  : entering actionCallCommitTransaction(), state: itx, actionNbr 0, nMsgs 1
2732.146409308:main Q:Reg/w0  : omfile: write to stream, pData->pStrm 0x7f25880022e0, lenBuf 117, strt data Jan 11 09:32:12 nyvm1131 rsyslogd-2007: can not initialize redis handle [v8.15.0 try http://www.rsyslog.com/e/2007 ]

2732.146430181:main Q:Reg/w0  : strm 0x7f25880022e0: file 11(messages) flush, buflen 117
2732.146443485:main Q:Reg/w0  : strmPhysWrite, stream 0x7f25880022e0, len 117
2732.146506319:main Q:Reg/w0  : strm 0x7f25880022e0: file 11 write wrote 117 bytes
2732.146518153:main Q:Reg/w0  : Action 0 transitioned to state: rdy
2732.146527916:main Q:Reg/w0  : Action 0 transitioned to state: itx
2732.146536653:main Q:Reg/w0  : Action 0 transitioned to state: rdy
2732.146545177:main Q:Reg/w0  : actionCommit, in retry loop, iRet 0
2732.146554524:main Q:Reg/w0  : processBATCH: batch of 1 elements has been processed
2732.146564759:main Q:Reg/w0  : regular consumer finished, iret=0, szlog 0 sz phys 1
2732.146579458:main Q:Reg/w0  : DeleteProcessedBatch: we deleted 1 objects and enqueued 0 objects
2732.146589259:main Q:Reg/w0  : doDeleteBatch: delete batch from store, new sizes: log 0, phys 0
2732.146599140:main Q:Reg/w0  : regular consumer finished, iret=4, szlog 0 sz phys 0
2732.146607985:main Q:Reg/w0  : main Q:Reg/w0: worker IDLE, waiting for work.
2732.147647036:remote:Reg/w0  : main Q: EnqueueMsg advised worker start
rsyslogd: can not initialize redis handle [v8.15.0 try http://www.rsyslog.com/e/2007 ]
2732.147696498:remote:Reg/w0  : Action 8 transitioned to state: rtry
2732.147708957:remote:Reg/w0  : actionDoRetry: redis2050-fwd enter loop, iRetries=0
2732.147719358:remote:Reg/w0  : actionDoRetry: redis2050-fwd action->tryResume returned 0
2732.147729670:remote:Reg/w0  : actionDoRetry: redis2050-fwd had success RDY again (iRet=0)
2732.148073981:remote:Reg/w0  : Called LogMsg, msg: action 'redis2050-fwd' resumed (module 'omhiredis')
2732.148112652:remote:Reg/w0  : main Q: qqueueAdd: entry added, size now log 1, phys 1 entries
2732.148142977:main Q:Reg/w0  : wti 0x7f25a0717080: worker awoke from idle processing
2732.148155562:main Q:Reg/w0  : DeleteProcessedBatch: we deleted 0 objects and enqueued 0 objects
2732.148164644:main Q:Reg/w0  : doDeleteBatch: delete batch from store, new sizes: log 1, phys 1
2732.148176652:main Q:Reg/w0  : processBATCH: batch of 1 elements must be processed
2732.148186604:main Q:Reg/w0  : processBATCH: next msg 0: action 'redis2050-fwd' resumed (module 'omhiredis') [v8.15.0 try http://www.rsyslog.com/e/2359 ]
2732.148198089:main Q:Reg/w0  :     PRIFILT '*.info;mail.none;authpriv.none;cron.none'
2732.148215297:main Q:Reg/w0  :     pmask: 7F 7F  X 7F 7F 7F 7F 7F 7F  X  X 7F 7F 7F 7F 7F 7F 7F 7F 7F 7F 7F 7F 7F 7F 7F
2732.148326548:main Q:Reg/w0  : PRIFILT condition result is 1
2732.148335259:main Q:Reg/w0  :     ACTION 0 [builtin:omfile:/var/log/messages]
2732.148353567:main Q:Reg/w0  : executing action 0
2732.148364099:main Q:Reg/w0  : action 'action 0': called, logging to builtin:omfile (susp 0/0, direct q 1)
2732.148379266:main Q:Reg/w0  : action 'action 0': is transactional - executing in commit phase
2732.148390921:main Q:Reg/w0  : Action 0 transitioned to state: itx
2732.148400489:main Q:Reg/w0  : action 'action 0': set suspended state to 0
2732.148409795:main Q:Reg/w0  :     PRIFILT 'authpriv.*'
2732.148425126:main Q:Reg/w0  :     pmask:  X  X  X  X  X  X  X  X  X  X FF  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X
2732.148527420:main Q:Reg/w0  : PRIFILT condition result is 0
2732.148535981:main Q:Reg/w0  :     PRIFILT 'authpriv.*'
2732.148551308:main Q:Reg/w0  :     pmask:  X  X  X  X  X  X  X  X  X  X FF  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X
2732.148653033:main Q:Reg/w0  : PRIFILT condition result is 0
2732.148661658:main Q:Reg/w0  :     PRIFILT 'mail.*'
2732.148676488:main Q:Reg/w0  :     pmask:  X  X FF  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X
2732.149242156:main Q:Reg/w0  : PRIFILT condition result is 0
2732.149245431:main Q:Reg/w0  :     PRIFILT 'cron.*'
2732.149250233:main Q:Reg/w0  :     pmask:  X  X  X  X  X  X  X  X  X FF  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X
2732.149282266:main Q:Reg/w0  : PRIFILT condition result is 0
2732.149292460:main Q:Reg/w0  :     PRIFILT '*.emerg'
2732.149297733:main Q:Reg/w0  :     pmask:  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1
2732.149332706:main Q:Reg/w0  : PRIFILT condition result is 0
2732.149335353:main Q:Reg/w0  :     PRIFILT 'uucp,news.crit'
2732.149340019:main Q:Reg/w0  :     pmask:  X  X  X  X  X  X  X  7  7  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X
2732.149371808:main Q:Reg/w0  : PRIFILT condition result is 0
2732.149374416:main Q:Reg/w0  :     PRIFILT 'local7.*'
2732.149379008:main Q:Reg/w0  :     pmask:  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X FF  X  X
2732.149410696:main Q:Reg/w0  : PRIFILT condition result is 0
2732.149413517:main Q:Reg/w0  : END batch execution phase, entering to commit phase
2732.149417412:main Q:Reg/w0  : actionCommitAll: action 0, state 1, nbr to commit 1 isTransactional 1
2732.149421324:main Q:Reg/w0  : doTransaction: have commitTransaction IF, using that, pWrkrInfo 0x7f25a0717150
2732.149425223:main Q:Reg/w0  : entering actionCallCommitTransaction(), state: itx, actionNbr 0, nMsgs 1
2732.149429215:main Q:Reg/w0  : omfile: write to stream, pData->pStrm 0x7f25880022e0, lenBuf 137, strt data Jan 11 09:32:12 nyvm1131 rsyslogd-2359: action 'redis2050-fwd' resumed (module 'omhiredis') [v8.15.0 try http://www.rsyslog.com/
2732.149435151:main Q:Reg/w0  : strm 0x7f25880022e0: file 11(messages) flush, buflen 137
2732.149438797:main Q:Reg/w0  : strmPhysWrite, stream 0x7f25880022e0, len 137
2732.149450034:main Q:Reg/w0  : strm 0x7f25880022e0: file 11 write wrote 137 bytes
2732.149453669:main Q:Reg/w0  : Action 0 transitioned to state: rdy
2732.149456778:main Q:Reg/w0  : Action 0 transitioned to state: itx
2732.149459462:main Q:Reg/w0  : Action 0 transitioned to state: rdy
2732.149462051:main Q:Reg/w0  : actionCommit, in retry loop, iRet 0
2732.149464666:main Q:Reg/w0  : processBATCH: batch of 1 elements has been processed
2732.149467926:main Q:Reg/w0  : regular consumer finished, iret=0, szlog 0 sz phys 1
2732.149472448:main Q:Reg/w0  : DeleteProcessedBatch: we deleted 1 objects and enqueued 0 objects
2732.149475628:main Q:Reg/w0  : doDeleteBatch: delete batch from store, new sizes: log 0, phys 0
2732.149478818:main Q:Reg/w0  : regular consumer finished, iret=4, szlog 0 sz phys 0
2732.149481486:main Q:Reg/w0  : main Q:Reg/w0: worker IDLE, waiting for work.
2732.149486551:remote:Reg/w0  : main Q: EnqueueMsg advised worker start
2732.149490670:remote:Reg/w0  : Action 8 transitioned to state: rdy
2732.149494089:remote:Reg/w0  : omhiredis: beginTransaction called
2732.149496849:remote:Reg/w0  : Action 8 transitioned to state: itx
2732.149499978:remote:Reg/w0  : entering actionCalldoAction(), state: itx, actionNbr 8
2732.149514848:remote:Reg/w0  : Called LogMsg, msg: action 'redis2050-fwd' resumed (module 'omhiredis')
2732.149523158:remote:Reg/w0  : main Q: qqueueAdd: entry added, size now log 1, phys 1 entries
2732.149528463:remote:Reg/w0  : main Q: EnqueueMsg advised worker start
2732.149531910:remote:Reg/w0  : omhiredis: endTransaction called
Segmentation fault

**Comment by zacharyzhao** [comment]
Do you have the core dump file? Its backtrace information may do some help.

**Comment by missnebun** [comment]
Sorry I don't have it.

**Comment by rgerhards** [comment]
@taotetek could you handle this?

**Comment by theoryno3** [comment]
@rgerhards @taotetek Just wondering about the status of this issue? I'm encountering similar segfaults but with omelasticsearch. Thanks!

**Comment by davidelang** [comment]
@theoryno3 are you having segfaults only if ES is not reachable? or are you getting some data to ES but occasionally having segfaults?

**Comment by pemontto** [comment]
We're also seeing this issue pushing data to Redis.

**Comment by bobthemighty** [comment]
@pemontto I've opened a new PR https://github.com/rsyslog/rsyslog/pull/1109

I *think* that should solve your problem. If you want retry, you will need to configure a queue and a retry count, like this:

```
       action(
	   # retry forever if redis goes away
            action.resumeretrycount="-1"
            # wait 5 secs between connects - this value doubles every 10 attempts.
	    action.resumeinterval="5"
            # set the spool location for the disk-assisted queue
            queue.spoolDirectory="/var/spool/rsyslog"
            # use a max 1G of disk for queued messages
	    queue.maxDiskSpace="1G"
            queue.filename="things-waiting-for-redis" 
            # setting a nice large queue size so we don't drop messages if redis goes away
            queue.size="1000000" 
            queue.type="LinkedList"
            type="omhiredis"

            mode="..."
            key="..."
            server="..."
            template="..."
        )
```

**Comment by pemontto** [comment]
Thanks @bobthemighty! I like the idea of Rsyslog handling the retries instead of the module itself.

**Comment by mostolog** [comment]
Couldn't this be closed in favor of #1109 ?
